src/adapter/tests.rs

```diff
@@ -99,6 +99,69 @@ fn impl_for_ref() {
     );
 }
 
+#[test]
+fn impl_name_for_impl_owners() {
+    let path = "./localdata/test_data/impl_for_ref/rustdoc.json";
+    let content = std::fs::read_to_string(path)
+        .with_context(|| format!("Could not load {path} file, did you forget to run ./scripts/regenerate_test_rustdocs.sh ?"))
+        .expect("failed to load rustdoc");
+
+    let crate_ = serde_json::from_str(&content).expect("failed to parse rustdoc");
+    let indexed_crate = IndexedCrate::new(&crate_);
+    let adapter = RustdocAdapter::new(&indexed_crate, None);
```

Comment:
Note how only rustdoc from `impl_for_ref` is included in the adapter. No `dummy_ext` rustdoc is loaded, so it's as if none of its types exist from the perspective of the `RustdocAdapter`.

Reply:
I don't know the internals but that's not entirely true.  

I have two reference to the external trait in the rustdoc json generated for `impl_for_ref`: 
```json
{
  "0:8": {
    "id": "0:8",
    "crate_id": 0,
    "name": null,
    "span": {
      "filename": "src/lib.rs",
      "begin": [
        9,
        0
      ],
      "end": [
        11,
        1
      ]
    },
    "visibility": "default",
    "docs": null,
    "links": {},
    "attrs": [],
    "deprecation": null,
    "inner": {
      "impl": {
        "is_unsafe": false,
        "generics": {
          "params": [],
          "where_predicates": []
        },
        "provided_trait_methods": [],
        "trait": {
          "name": "DummyExternalTrait",
          "id": "20:3:2305",
          "args": {
            "angle_bracketed": {
              "args": [],
              "bindings": []
            }
          }
        },
        "for": {
          "resolved_path": {
            "name": "StringHolder",
            "id": "0:5:2306",
            "args": {
              "angle_bracketed": {
                "args": [],
                "bindings": []
              }
            }
          }
        },
        "items": [
          "0:9:2308"
        ],
        "negative": false,
        "synthetic": false,
        "blanket_impl": null
      }
    }
  }
}
```
```json
{
  "20:3:2305": {
    "crate_id": 20,
    "path": [
      "dummy_ext",
      "DummyExternalTrait"
    ],
    "kind": "trait"
  }
}
```

Reply:
> Unfortunately this wouldn't help. 

I am puzzled. I turned to whole thing into a static mutable, add added this to the test and now it passes. 
```rust
    crate::indexed_crate::add_manual_trait_item(ManualTraitItem {
        name: "DummyExternalTrait",
        is_auto: false,
        is_unsafe: false,
    });
```

Reply:
What I mean is that the actual definition of item `20:3:2305` AKA `DummyExternalTrait` is not present in the file: we don't know its attributes, its doc comment, its items, etc. We just know it as a foreign item from an _unknown version_ of a crate called `dummy_ext`.

With the approach you propose, queries like the following will produce unexpected results: [playground](https://play.predr.ag/rustdoc#?f=2&q=query---0Crate---2item---4*E-ImplOwner---6name-*o*l--_6impl---8implemented_trait---aname-*o*9*n-**impls***0*l*l--_aimpls_*B-trait---cdocs-*o*l--_cdoc_hidden-*o*l*l--_cimportable_path---epath-*o*l--_c--*a--*8--*6--*4--*2--*0*J*l*J&v=*C*l*l*J)
```graphql
query {
  Crate {
    item {
      ... on ImplOwner {
        name @output
        impl {
          implemented_trait {
            name @output(name: "impls")

            impls_: trait {
              docs @output  # <-- no data here
              doc_hidden @output  # <-- no data here either

              importable_path {  # <-- no known instances of this edge
                path @output
              }
            }
          }
        }
      }
    }
  }
}
```
It will claim that the implemented `DummyExternalTrait` trait is not actually importable from other crates (obviously false — it's implemented by a type in another crate!), because the data to determine its import paths was not present.

That sounds like a pretty serious footgun to me — one made even more unpredictable and dangerous if we add `add_manual_trait_item()` to the crate's public API. This is why I'd love to hear more about your use case — there may be a less footgun-y way to accomplish what you're after.

Reply:
>This is why I'd love to hear more about your use case 

I am trying to generate automated documentation for the zed editor actions (those are struct implementing the `Action` trait spread across several crates in a cargo workspace).  
I could probably parse everything with syn but I was curious about trustfall.

Reply:
In that use case, is it guaranteed that you'll have at most one major version of each crate you are querying in the workspace? (This isn't true in general, for example it isn't true when semver-checking `cargo-semver-checks`.)

If that's the case, you should be able to accomplish the goal with Trustfall in a better (but slightly more work) way: you could generate the rustdoc for all the crates, and use the external item description
```
{
  "20:3:2305": {
    "crate_id": 20,
    "path": [
      "dummy_ext",
      "DummyExternalTrait"
    ],
    "kind": "trait"
  }
}
```
to then look up the item in the other rustdoc using [the import path index](https://github.com/obi1kenobi/trustfall-rustdoc-adapter/blob/rustdoc-v29/src/indexed_crate.rs#L22-L23): crate `dummy_ext` import path `dummy_ext::DummyExternalTrait`.

Such a prototype would lay the groundwork for `trustfall-rustdoc-adapter` supporting cross-crate querying once the compiler MCP is approved and the rustdoc linking limitation is lifted. I'd love to see it and if you're interested in building it, I'd love to support you in it.

But I totally understand if that sounds like too much work, and you'd prefer to just fork the crate and make the `add_manual_trait_item()` function public in your build.

Reply:
I will give a try, this sound like a lot of fun! I'll let you know if I need your help later.

 Thanks a lot for the help so far !

Reply:
Awesome! Always happy to help :)

Reply:
Hi! I wanted to check in and see how you're doing on this. Are you still working on the prototype, or did something else take priority? Anything I can do to help you along?